### PR TITLE
DRILL-8358: Storage plugin for querying other Apache Drill clusters

### PR DESCRIPTION
**Pull request by vvysotskyi**
# [DRILL-8358](https://issues.apache.org/jira/browse/DRILL-8358): Storage plugin for querying other Apache Drill clusters

## Description
Using native client to query other drill clusters. Added logic to do various pushdowns when possible.
Fixed adding extra project for the case of star columns.
Fixed ignoring column with empty name column for excel format.

## Documentation
See README.md

## Testing
Tested manually, added UT.

### CODE REVIEW

File: contrib/storage-drill/src/main/java/org/apache/drill/exec/store/drill/plugin/DrillStoragePluginConfig.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  public DrillStoragePluginConfig updateCredentialProvider(CredentialsProvider credentialsProvider) {
+    return new DrillStoragePluginConfig(this, credentialsProvider);
+  }
+
+  @JsonIgnore
+  public DrillClient getDrillClient(String userName, BufferAllocator allocator) {
+    try {
+      String urlSuffix = connection.substring(CONNECTION_STRING_PREFIX.length());
+      Properties props = ConnectStringParser.parse(urlSuffix, properties);
+      props.putAll(credentialsProvider.getUserCredentials(userName));
```
Comment (jnturton) [inline contrib/storage-drill/src/main/java/org/apache/drill/exec/store/drill/plugin/DrillStoragePluginConfig.java:96]: This getUserCredentials(String username) method is meant to fetch per-query-user credentials for plugins that are in user translation auth mode while the nullary method getUserCredentials() is meant for shared credentials. Only the plan and Vault providers currently support per-user credentials. You can see some logic for deciding which to call  (via UsernamePasswordCredentials objects) in JdbcStorageConfig on line 142.

Those APIs wound up being a little ugly :/

**Review by jnturton — CHANGES_REQUESTED** [review]

File: contrib/storage-drill/src/main/java/org/apache/drill/exec/store/drill/plugin/DrillStoragePluginConfig.java (hunk 1 above)
Comment (vvysotskyi) [reply contrib/storage-drill/src/main/java/org/apache/drill/exec/store/drill/plugin/DrillStoragePluginConfig.java:96]: Thanks, fixed.

**Review by cgivre — COMMENTED** [review]
Thanks @vvysotskyi  for this.  I had a usage question as well. 

Let's say that I have 2 drills, drill1 and drill2.  Let's say that drill2 is connected to a file system called dfs2 and I want to query that from drill1.   What would the query look like?
Would it be something like:

```sql
SELECT *
FROM drill2.dfs.ws.`file`
```

File: contrib/storage-drill/src/main/java/org/apache/drill/exec/store/drill/plugin/DrillSubScan.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  private final String query;
+
+  @JsonProperty
+  private final DrillStoragePluginConfig pluginConfig;
+
+  @JsonCreator
+  public DrillSubScan(
+      @JsonProperty("userName") String userName,
+      @JsonProperty("mongoPluginConfig") StoragePluginConfig pluginConfig,
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply contrib/storage-drill/src/main/java/org/apache/drill/exec/store/drill/plugin/DrillSubScan.java:47]: No, it isn't, thanks, fixed it.

**Comment by vvysotskyi** [comment]
@cgivre, yes, you can create a plugin in drill1 with the name drill2, and query all plugins that drill2 has configured from drill1, so if drill2 has file system plugin called dfs2, query for drill1 will be the following:
```sql
SELECT *
FROM drill2.dfs2.ws.`file`
```

**Pull request merged**